pkg/client/client_venafi_cloud.go

```diff
@@ -0,0 +1,377 @@
+package client
+
+import (
+	"bytes"
+	"crypto"
+	"crypto/ecdsa"
+	"crypto/ed25519"
+	"crypto/rsa"
+	"crypto/x509"
+	"encoding/json"
+	"encoding/pem"
+	"fmt"
+	"io"
+	"io/ioutil"
+	"net/http"
+	"net/url"
+	"os"
+	"path"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"sync"
+	"time"
+
+	"github.com/golang-jwt/jwt/v4"
+	"github.com/google/uuid"
+	"github.com/hashicorp/go-multierror"
+	"github.com/jetstack/preflight/api"
+	"gopkg.in/yaml.v2"
+)
+
+type (
+	// The VenafiCloudClient type is a Client implementation used to upload data readings to the Venafi Cloud platform
+	// using service account authentication as its authentication method.
+	//
+	// This form of authentication follows the Private Key JWT standard found at https://oauth.net/private-key-jwt,
+	// which is a combination of two RFCs:
+	// * RFC 7521 (Assertion Framework)
+	// * RFC 7523 (JWT Profile for Client Authentication)
+	VenafiCloudClient struct {
+		credentials   *VenafiSvcAccountCredentials
+		accessToken   *venafiCloudAccessToken
+		baseURL       string
+		agentMetadata *api.AgentMetadata
+		client        *http.Client
+
+		uploadID      string
+		uploadPath    string
+		privateKey    crypto.PrivateKey
+		jwtSigningAlg jwt.SigningMethod
+		lock          sync.RWMutex
+	}
+
+	VenafiSvcAccountCredentials struct {
+		// ClientID is the service account client ID
+		ClientID string `yaml:"client_id,omitempty"`
```

Comment:
Question: was there a reason for using yaml instead of json for the credentials?

Reply:
I think because that is the format of the credentials from Venafi cloud, but let's wait for Andrey to confirm

Reply:
Yes, it was copied from another way to read in those credentials. Will change to json to match OAuth.